### ISSUE TITLE
Support direct HTTP retrieval from /https providers

### ISSUE THREAD

**Issue by lidel** [ipfs/rainbow#125]
This is GO version of https://github.com/ipfs-shipyard/service-worker-gateway/issues/72.

We want rainbow to benefit from `/https` providers ([example](https://delegated-ipfs.dev/routing/v1/providers/bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi)) and use them in addition to bitswap 

Ideally, we would be prioritizing HTTP retrieval over bitswap, where possible, as it lowers the cost of content providers, and incentivizes them to configure, expose, and announce HTTPS endpoints.

## MVP scope

Focus should be on block ([application/vnd.ipld.raw](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw), `?format=raw`) requests, as these will always work, across all implementations, and provide the best cachability for HTTP infrastructure we have.

[CAR with IPIP-402](https://specs.ipfs.tech/ipips/ipip-0402/) may be more involved, and may lead to duplicated block retrievals due to the way loading a page with a dozen of subresources works (all share the same parent, all fetched in parallel, may lead to racy case where parent blocks are fetched multiple times, slowing down page loads)

**Comment by hacdias** [comment]
Before continuing, I want to lay down some notes to make sure we're all on the same page about what needs to be done and about the current challenges with accepting the `/https` providers.

## Most providers with HTTPS multiaddresses are unusable

Most, if not all, providers advertising `/https` multiaddresses are, standard-speaking, unusable. They do not follow the proper [peer schema](https://specs.ipfs.tech/routing/http-routing-v1/#peer-schema). We can certainly hammer the code to accept them, but I would rather have the original provider of the records implement the correct schema instead. So, instead of:

```json
{
  "Addrs": ["/dns4/dag.w3s.link/tcp/443/https"],
  "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
  "Metadata": "oBIA",
  "Protocol": "transport-ipfs-gateway-http",
  "Schema": "unknown"
},
```

We should be getting this:

```json
{
  "Schema": "peer",
  "ID": "QmUA9D3H7HeCYsirB3KmPSvZh3dNXMZas6Lwgr4fv1HTTp",
  "Addrs": ["/dns4/dag.w3s.link/tcp/443/https"],
  "Protocols": ["transport-ipfs-gateway-http"]
}
```

As I said, the code can be hammered to accept this (albeit a bit harder in Go). But I would rather not go that avenue. We already have plans of completely removing support for `"Schema": "bitswap"` (e.g.: from Pinata) from [Boxo](https://github.com/ipfs/boxo/pull/539). Supporting one more non-standardized schema will just make things more complicated when it doesn't need to be.

## Fetching the block via HTTPS

The current flow to fetch a block, from the Blockservice perspective, is as follows:

1. Blockservice gets asked for a block
2. Blockservice checks with Blockstore, if it has it, return it. Otherwise,
3. Blockservice asks the Exchange, which currently is just Bitswap
4. Bitswap looks out for providers using a `routing.ContentRouter`. This `routing.ContentRouter` only has Bitswap-related peers. All other peers are ignored, even if they come from a [`/routing/v1` endpoint](https://github.com/ipfs/boxo/blob/eeea414587350401b6b804f0574ed8436833331d/routing/http/contentrouter/contentrouter.go#L60).
5. Bitswap tries fetching it, returns, etc, etc.

I see a few ways of potentially solving this.

### (a) Parallel Exchanges

Create a parallel exchange that calls both Bitswap and a new exchange that can take advantage of the Delegated Routing endpoint results that have non-Bitswap peers. This will lead to duplicate HTTP requests to the delegated routing endpoints, done by both exchanges. This is not ideal.

### (b) Smarter Exchange

A smarter exchange that calls the routing itself and, depending on the results, tries using Bitswap to fetch Bitswap things, or HTTP requests for the other peers that have such addresses. The challenges I see here:

1. We need from the exchange to be able to directly tell Bitswap that we know that peer _X_ has the block _Y_. Maybe this is already possible, but I'm not familiar with the code and would have to better investigate.
2. See how to reconcilliate the Delegated Routing lookups with DHT lookups. This is already done in [someguy](https://github.com/ipfs-shipyard/someguy), which parallelizes DHT and Delegated Routing endpoints into a Delegated Routing-like interface. Thus, this is the inverse of what Boxo does (Delegated Routing -> Libp2p Routers).

**Comment by lidel** [comment]
Triage: 
- Most providers with HTTPS multiaddresses are unusable
  - try to clean up `cid.contact/routing/v1` responses, switch both http and bitswap providers to modern `peer` schema so we can remove hacks from boxo/kubo/rainbow
- look into Smarter Exchange
  - assumption: we always have _some_ peerid and some Addrs, so we cna reuse interfaces from libp2p
  - try to libp2p identify to learn p2p protocols, and use bitswap if present
  - if `/http/tls` or `/https` or `/http` is present, attempt HTTP retrieval instead of bitswap

**Comment by hacdias** [comment]
Update:

* `cid.contact/routing/v1` now returns records with the peer schema

**Comment by 2color** [comment]
This issue can be closed now that  https://github.com/ipfs/rainbow/pull/242 is merged.
